### PR TITLE
[Forward Port PR-14344] Fix generating product URL rewrites for anchor categories

### PR DESCRIPTION
**Pull request by hostep**
(cherry picked from commit 63111ed399b2e058efc0bf1c7c5427299cb8c5fc)

### Description (*)
This is a forward port of https://github.com/magento/magento2/pull/14344 to 2.4-develop
This replaces https://github.com/magento/magento2/pull/20826 which is stuck because the author is no longer responding for a few months.
Previous attempts at fixing this issue were done in https://github.com/magento/magento2/pull/7667 & https://github.com/magento/magento2/pull/22212

**FOR THE MAINTAINERS**: if this gets approved, can you please please please backport this to Magento 2.3, because the change is already in Magento 2.2, and by this PR also in 2.4, so it should also be part of 2.3, so we have consistent behavior over all these major Magento versions.

### Related Pull Requests
- https://github.com/magento/magento2/pull/7667
- https://github.com/magento/magento2/pull/14344
- https://github.com/magento/magento2/pull/20826
- https://github.com/magento/magento2/pull/22212

### Fixed Issues (if relevant)
1. https://github.com/magento/magento2/issues/4112: Wrong parent category url_key in URL
2. https://github.com/magento/magento2/issues/11615: URL Rewrites vs multiple storeviews - a never ending battle
3. https://github.com/magento/magento2/issues/11616: URL Rewrites vs multiple storeviews - too many rewrites are being generated
4. https://github.com/magento/magento2/issues/25124: Magento 2.3 Wrong product url for anchor categories for multiple storeviews
5. https://github.com/magento/magento2/issues/26393: Product category url rewrite missing storeview specific url_key
6. Probably some more ...

### Manual testing scenarios (*)
1. See steps to reproduce in https://github.com/magento/magento2/issues/11615 and https://github.com/magento/magento2/issues/11616

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by rogyar** [comment]
Hi @hostep. It would be awesome to have this fix finally merged. But considering the number of talks/attempts, it will be fair to cover this fix using a functional test. Could I kindly ask you to do that? 

Thank you!

**Comment by hostep** [comment]
I agree @rogyar!

Due to me not having much time free and having very little experience with writing tests, somebody else may pick this up if they want.
Maybe I'll have some more time later this week, we'll see if it happens or not from my part.

**Comment by hostep** [comment]
Thanks for the test @engcom-Golf!

But I'm confused, since it doesn't seem to be testing the generated url rewrites of products belonging to categories of at least 2 levels deep.
Am I missing something obvious here?

Also: why is the parent_id, path & level of those categories with ids 3 & 4 different on a different storeview, that seems like a mistake?

**Comment by engcom-Echo** [comment]
Hi @hostep. Thanks for your comment.

The test checks this code if we use the parameters of $storeId.

> $this->categoryRepository->get($anchorCategoryId, $storeId);

These fixtures are a little scary but in my opinion, we should check the code for different use $storeId - that's exactly what I saw an important change.

**Comment by hostep** [comment]
Ok thanks for the feedback!

In my opinion, we should also test if the generated url_rewrites are correct, as that's the purpose of this PR, see https://github.com/magento/magento2/issues/11615 for what the correct values should be in this case.

Could we add a test for that as well? Thanks!

**Comment by engcom-Echo** [comment]
I will take care of functional test coverage.

**Comment by hostep** [comment]
Thanks for the MFTF test @engcom-Echo!

I think I understand how it works. But just to be clear: this test is failing on the code as it was before this fix, right?

And shouldn't we also test the `url_key_custom_store`? Or is that not necessary? Something like:
```xml
<amOnPage url="{{category.url_key_custom_store}}/{{subCategory.url_key_custom_store}}/{{product.urlKey}}2.html" stepKey="goToProductPage"/>
<see selector="{{StorefrontProductInfoMainSection.productName}}" userInput="{{product.name}}" stepKey="seeProductNameInStoreFront"/>
```
(I'm not sure how the url is build when you should be on a different store in MFTF though)

Thanks!

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/CatalogUrlRewrite/Test/Mftf/ActionGroup/StorefrontAssertProductRewriteUrlSubCategoryActionGroup.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontAssertProductRewriteUrlSubCategoryActionGroup">
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Mftf/ActionGroup/StorefrontAssertProductRewriteUrlSubCategoryActionGroup.xml]: Hi @engcom-Echo. Could I ask you to rename the action group in order to follow the general naming rules, please? So we have

`AssertStorefrontProductRewriteUrlSubCategoryActionGroup`

in the result.

Please, check the [technical guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html) (part 11.3.5).

Thank you.

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/ChangeSeoUrlKeyForSubCategoryWithoutRedirectActionGroup.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="ChangeSeoUrlKeyForSubCategoryWithoutRedirectActionGroup" extends="ChangeSeoUrlKeyForSubCategoryActionGroup">
```
Comment (rogyar) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/ChangeSeoUrlKeyForSubCategoryWithoutRedirectActionGroup.xml]: Please, use the area prefix (Admin|Storefront) for this action group. If we have all action groups prefixed correctly, it's much easier to find and reuse an action group in the future.

Thank you!

**Review by rogyar — DISMISSED** [review]

File: app/code/Magento/CatalogUrlRewrite/Test/Mftf/ActionGroup/StorefrontAssertProductRewriteUrlSubCategoryActionGroup.xml (hunk 1 above)
Comment (engcom-Echo) [reply app/code/Magento/CatalogUrlRewrite/Test/Mftf/ActionGroup/StorefrontAssertProductRewriteUrlSubCategoryActionGroup.xml]: Thank you

**Comment by engcom-Echo** [comment]
Hi @rogyar  could you please approve it? :)

**Review by rogyar — APPROVED** [review]
Thank you @engcom-Echo. You rock!

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-6997](https://jira.corp.magento.com/browse/ENGCOM-6997) has been created to process this Pull Request

File: dev/tests/integration/testsuite/Magento/CatalogUrlRewrite/Model/Product/AnchorUrlRewriteGeneratorTest.php [hunk 3]
```diff
@@ -50,32 +44,49 @@ public function setUp()
...
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
      * @magentoDataFixture Magento/CatalogUrlRewrite/_files/product_with_stores.php
      * @magentoDbIsolation disabled
-     *
-     * @return void
+     * @dataProvider getConfigGenerate
      */
-    public function testGenerate(): void
+    public function testGenerate(string $expect): void
```
Comment (lbajsarowicz) [inline dev/tests/integration/testsuite/Magento/CatalogUrlRewrite/Model/Product/AnchorUrlRewriteGeneratorTest.php:61]: For what input should the result change?
I see expectation attribute, but there's not context attribute.

**Comment by lbajsarowicz** [comment]
@mszydlo Today I wanted to forward this change, then realised that @hostep is doing it already for 21 days. Awesome work! I can't wait to get this merged.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed 

**Before:**
before

**After:**
after

**Comment by slavvka** [comment]
@magento run Functional Tests EE

**Comment by hostep** [comment]
Thanks @slavvka, do you know if this is going to be backported to Magento 2.3? Since https://github.com/magento/magento2/pull/14344 was merged in 2.2, we should probably also have this fix on 2.3, no?

**Comment by hostep** [comment]
@slavvka, @chernenm, can you guys let me know if this is being backported to the 2.3.x release?

We have this awkward situation where this bug is fixed in Magento 2.2.10, 2.2.11 and 2.4.0 (I guess?), but the fix is not in 2.3.x ...

**Comment by peterjaap** [comment]
@hostep seems like at least a part of this PR is present in 2.3.6; https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/CatalogUrlRewrite/Model/Product/AnchorUrlRewriteGenerator.php#L70

**Comment by hostep** [comment]
Ah nice find @peterjaap !

It was added in [MC-32810: [Magento Cloud] - Incorrect URL Rewrites for non-default websites after upgrade to 2.3.4](https://github.com/magento/magento2/commit/82c91868a2d47d9c052149da1908c5ce1903587b) (what a silly description, this has been a bug for many years)

However, it seems to consist of two changes and only one of them is done in 2.4-develop not the other.

Fix one: [2.3.6](https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/CatalogUrlRewrite/Model/CategoryProductUrlPathGenerator.php#L47) vs [2.4-develop](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogUrlRewrite/Model/CategoryProductUrlPathGenerator.php#L46-L48) => missing in 2.4
Fix two: [2.3.6](https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/CatalogUrlRewrite/Model/Product/AnchorUrlRewriteGenerator.php#L70) vs [2.4-develop](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogUrlRewrite/Model/Product/AnchorUrlRewriteGenerator.php#L70) => both in 2.3 and 2.4

So this is again a weird thing and maybe somebody forgot to do something here.

@slavvka, @chernenm, @sdzhepa: can we figure this out please? Was that first change in `MC-32810` forgotten to be included in `2.4-develop`, or was this left out intentionally for some reason?

**Pull request merged**